feat: Add namespace tags to FDB metrics

pboros opened the pull request:
Add namespace tags to FDB metrics. The GetInternalDatabase() call (used by CDC) will always be tagged as default_namepsace.

ovaistariq commented on server/metrics/fdb.go:
```diff
@@ -42,31 +45,53 @@ var (
-		"tx":     strconv.FormatBool(tx),
+		"method":    reqMethodName,
+		"tx":        strconv.FormatBool(tx),
+		"namespace": "default_namespace",
```
Should we shorten it to just default?

ovaistariq replied:
I also recommend defining it as a constant.

ovaistariq commented on server/metrics/fdb.go:
```diff
@@ -42,31 +45,53 @@ var (
 
+func addNameSpaceToTags(ctx context.Context, tags map[string]string) map[string]string {
+	namespace, err := request.GetNamespace(ctx)
+	if err != nil {
```
Combine the two if conditions

himank commented on server/request/metadata.go:
```diff
@@ -1,4 +1,4 @@
-package middleware
+package request
```
just call the file name request.go?

himank commented on server/services/v1/sessions.go:
```diff
@@ -16,6 +16,7 @@ package v1
 
 import (
 	"context"
+	"github.com/tigrisdata/tigris/server/request"
```
imports.

pboros replied:
Done.

himank commented on server/metrics/metrics.go:
```diff
@@ -27,6 +27,10 @@ import (
 )
 
+const (
+	DefaultReportedNameSpace string = "default_namespace"
```
can you move this in types.go inside the server/types package and let's call it `DefaultNamespace`

pboros replied:
This is done.

himank commented on server/metrics/fdb.go:
```diff
@@ -42,31 +45,51 @@ var (
 
+func addNameSpaceToTags(ctx context.Context, tags map[string]string) map[string]string {
+	namespace, err := request.GetNamespace(ctx)
+	if err != nil && err.Error() != "Namespace not found" {
```
can we construct a proper error when `namespace not found` and then use that for comparison here?

pboros replied:
This is done.

pboros commented:
I modified this PR based on the comments:
- Added specific error for namespace not found.
- If the namespace is not known, the tag will report "unknown"
- For the admin api endpoints, no namespace tag is added
- The GetInternalDatabase is no longer instrumented
- Addressed the code review comments like file naming, imports, etc

himank commented on server/midddleware/metrics.go:
```diff
@@ -33,22 +34,34 @@ import (
-		ulog.E(err)
+		switch err.(type) {
+		case *request.NamespaceNotFound:
+			tags["namespace"] = "unknown"
```
@JigarJoshi I guess you are already adding it in auth interceptor, do we need to do it here?

JigarJoshi replied:
this can be avoided here. auth namespace will place `unknown` if the auth fails and namespace is supplied in `""`

JigarJoshi replied:
in all other cases, there will always be a namespace.

pboros commented on server/metrics/fdb.go:
```diff
@@ -42,31 +45,53 @@ var (
 
+func addNameSpaceToTags(ctx context.Context, tags map[string]string) map[string]string {
+	namespace, err := request.GetNamespace(ctx)
+	if err != nil {
```
Done.

himank commented on server/metrics/fdb.go:
```diff
@@ -38,23 +42,42 @@ var (
+		"method":        reqMethodName,
+		"error_code":    code,
+		"tx":            strconv.FormatBool(tx),
+		"tigris_tenant": DefaultReportedTigrisTenant,
```
do you need it here as `addTigrisTenantToTags ` is setting it?

himank commented on server/metrics/fdb.go:
```diff
@@ -38,23 +42,42 @@ var (
-		"tx":     strconv.FormatBool(tx),
+		"method":        reqMethodName,
+		"tx":            strconv.FormatBool(tx),
+		"tigris_tenant": DefaultReportedTigrisTenant,
```
same, do we need it here?

pboros commented on server/request/request.go:
```diff
@@ -27,6 +26,18 @@ type NamespaceExtractor interface {
 type AccessTokenNamespaceExtractor struct {
 }
 
+type NamespaceNotFound struct {
```
Done.

pboros commented on server/midddleware/metrics.go:
```diff
@@ -33,22 +34,34 @@ import (
+		case *request.NamespaceNotFound:
+			tags["namespace"] = "unknown"
+		default:
+			ulog.E(err)
```
Done.

himank commented on server/metrics/fdb.go:
```diff
@@ -38,23 +42,42 @@ var (
-		"method": reqMethodName,
-		"tx":     strconv.FormatBool(tx),
+		"method":        reqMethodName,
+		"tx":            strconv.FormatBool(tx),
```
everything is transactional, do we need this extra tag for now? Or pass it in the method for completion but no need to emit it?

pboros commented on server/metrics/requests.go:
```diff
@@ -33,20 +37,38 @@ func newRequestEndpointMetadata(serviceName string, methodInfo grpc.MethodInfo)
+		return map[string]string{
+			"method":        g.methodInfo.Name,
+			"grpc_service":  g.serviceName,
+			"tigris_tenant": SystemTigrisTenantName,
```
Yes.
